WIP: Add middleware that normalizes subdomain vs. path url scheme.

asynts opened the pull request:
This is a first step for #22. I added some middleware that rewrites as follows:

~~~none
https://codidact.com         -> https://codidact.com
https://foo.codidact.com     -> https://codidact.com/community/foo/
https://foo.codidact.com/bar -> https://codidact.com/community/foo/bar
~~~

*The following has been edited.*

The following things still need to be addressed: (not in this pull request though)

 1. `foo.codidact.com` and `codidact.com/community/foo` are both accepted, maybe redirect
     to the subdomain version when enabled?

luap42 commented:
I can't comment on the other stuff, but the first:

> foo.codidact.com and codidact.com/community/foo are both accepted, maybe redirect
to the subdomain version when enabled?
> This would be nice, but very difficult to implement because it is unclear whether
codidact.local/community/foo/bar should redirect to foo.codidact.local/bar or
bar.foo.codidact.local.

IMO, yes they should redirect to the "canonical" sub domain url, when a site setting is enabled for that.

I think only the first "section" after "/community/" should redirect. It should be reasonably limited in character type and count (`[a-z0-9-]{1,40}`), and everything else is part of the real URL (/question/42/why-is-codidact-so-amazing)

asynts commented:
That would work, but it would become a bit weird if we allow e.g. `wiki.foo.codidact.com` then `codidact.com/foo/wiki` would redirect to `foo.codidact.com/wiki` (If I understand you correctly) which is a bit weird.

Notice, that `foo.codidact.com/wiki` would work either way because it is understood as `codidact.com/foo/wiki` internally. (Maybe we should redirect that too?)

asynts committed:
Add middleware that normalizes the url schema.

Refer to https://forum.codidact.org/t/what-should-be-the-url-pattern-for-communities/650,
a few examples:

https://codidact.com         -> https://codidact.com
https://foo.codidact.com     -> https://codidact.com/community/foo/
https://foo.codidact.com/bar -> https://codidact.com/community/foo/bar

luap42 commented:
Well then redirect the second part, too, if it is one of a list of possible subdomains (only "meta" for now)

asynts commented:
That's exactly the thing that I described as "difficult". If we do something like that, it could bite us later, when we e.g. add a third level subdomain, e.g. `edit.wiki.foo.codidact.com`. Then we would need some sort of tree structure which has to be maintained along the controller code.

I think we should make a decision to only allow a maximum of two levels of redirection, or maybe even only one. And then do it like you suggested.

luap42 commented:
Considering this again, I think that Monica's post summarizes this quite well:

> > There’s actually a good reason to put the wiki. / meta. sections further right in the domain.
>
> Also, things “within” a site belong to the right; meta or wiki or whatever isn’t a whole separate site but a sub-part, so it shouldn’t have its own URL. (Yes this is a change from how SE handles meta; we’re not going to have a whole separate site but, rather, a tab or something for those questions.)

So at maximum one level for the community.

asynts commented:
~~I just noticed that there is existing middleware for that, we should probably use that.~~ It's not powerful enough.

[URL Rewriting Middleware in ASP.NET Core][1]

  [1]: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/url-rewriting?view=aspnetcore-3.1

misha130 commented:
Really dont understand your end goal here. Do you want to route or to redirect? Redirecting should probably be done in the server configuration and not so much in code imo.

asynts commented:
There are two things supposed to happen:

 1. Rewrite e.g. `foo.codidact.com/meta` to `codidact.com/community/foo/meta` **internally**; this would not
    change what the user sees in the browser. This way both the subdomain schema and path schema
    are treated equally by the implementation.

 2. Redirect the path schema to the subdomain schema when it is enabled. Thus the user would see
    `foo.codicat.com/meta` (which would internally be handled as `codidact.com/community/foo/meta`) even if he
    explicitly types `codicat.com/community/foo/meta`.

luap42 commented:
@asynts I'd, however, not have codidact.com/foo/meta, but codidact.com/community/foo/meta (or any other - ideally configurable - separator)

asynts commented:
That was a typo. I edited the post.

~~**Addendum.** Here is my [current status][1] if someone want's to pick that up. I have to get up early tomorrow and it's getting late in UTC+1.~~ [current status][2]

  [1]: https://github.com/asynts/codidact-core/commits/asynts.temporary.1
  [2]: https://github.com/asynts/codidact-core/commits/temporary.2

asynts commented:
There is another thing to be discussed, should `codidact.com/comunidad/foo` be rewritten to `codidact.com/community/foo`? ("comunidad" is Spanish and could be another value for a community separator)

I'll try to finish a smaller version that only does the url schema rewrite, and then we should discuss this in the forum.

ArtOfCode- commented:
> There is another thing to be discussed, should codidact.com/comunidad/foo be rewritten to codidact.com/community/foo?

No, at least not for now. We'll want to revisit that as and when we introduce localization, but for now that section of the URL should just be a single configurable "community separator" value.

asynts commented:
I've now reduced the scope and only implemented the rewriting stuff. I edited the pull request to list the things that need to be addressed separately.

misha130 commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,47 @@
+using System;
```
Move this file to folder called Middlewares in WebUI

asynts replied:
I first moved it into `Middleware`, I changed this now to `Middlewares`.

misha130 commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,47 @@
+                var path = string.Join("/", parts.Reverse());
+
+                context.Request.Host = new HostString(options.Value.Hostname);
+                context.Request.Path = $"/community/{path}{context.Request.Path}";
```
Shouldnt "community" be the actual subdomain?

asynts replied:
A request to `foo.codidact.com` should be rewritten to `codidact.com/community/foo` not `codidact.com/foo`. [This was the consensus.][1] This should be correct.

  [1]: https://forum.codidact.org/t/what-should-be-the-url-pattern-for-communities/650/37?u=asynts

misha130 commented:
Can we get some tests going for all the options?

ranolfi commented:
> Note: this pull request is not ready to merge, there are still a few things to decide.

Since this is still in discussion phase, I added `WIP:` to the title. In the future please open a draft PR instead. Refer to: https://github.blog/2019-02-14-introducing-draft-pull-requests/. Thanks!

asynts committed:
Add integration tests for url schema middleware.

asynts commented:
I rebased my changes, cleaning up the commit history. The pull request now includes the following:

 1. Rewrite subdomain schema to path schema, e.g.
    ~~~none
    https://foo.codidact.com/meta -> https://codidact.com/foo/meta
    ~~~
    This can be configured in the `appsettings.json` file.

 2. Rewrite invalid requests with multiple subdomains to `/error/404`, e.g.
    ~~~none
    https://a.b.codidact.com/meta -> https://codidact.com/error/404
    ~~~

 3. The community separator can be configured and will be rewritten to "community", e.g.
    ~~~none
    https://codidact.com/comunidad/foo -> https://codidact.com/community/foo
    ~~~

The pull request **doesn't** include the following features, they are subject to another pull request:

 1. Redirect the path schema to the subdomain schema if enabled. (Please note: "redirect" not "rewrite".) e.g.
    ~~~none
    https://codidact.com/community/foo -> https://foo.codidact.com/`
    ~~~

 2. Many invalid requests are still accepted, e.g.
    ~~~none
    https://codidact.com/community
    ~~~

**[Cellio set a deadline here, giving people time to dispute the current url schema. This pull request should not be merged until that deadline runs out.][1]**

  [1]: https://forum.codidact.org/t/what-should-be-the-url-pattern-for-communities/650/37?u=asynts

ArtOfCode- commented on src/Domain/Extensions/StringExtensions.cs:
```diff
@@ -18,5 +18,13 @@ public static string ToSnakeCase(this string input)
         }
+
+        public static string TrimSuffix(this string input, string suffix) {
+            if(input.EndsWith(suffix)) {
```
Missing space between `if` and opening parenthesis: `if (input.EndsWith(suffix))`. Missing newline before opening brace:

```c#
if (input.EndsWith(suffix))
{
```

ArtOfCode- commented on src/Domain/Extensions/StringExtensions.cs:
```diff
@@ -18,5 +18,13 @@ public static string ToSnakeCase(this string input)
+        public static string TrimSuffix(this string input, string suffix) {
+            if(input.EndsWith(suffix)) {
+                return input.Substring(0, input.Length - suffix.Length);
+            } else {
```
New line after terminating brace, new line before opening brace:

```c#
}
else
{
```

ArtOfCode- commented on src/Domain/Extensions/StringExtensions.cs:
```diff
@@ -18,5 +18,13 @@ public static string ToSnakeCase(this string input)
             return startUnderscores + Regex.Replace(input, @"([a-z0-9])([A-Z])", "$1_$2").ToLower();
         }
+
+        public static string TrimSuffix(this string input, string suffix) {
```
Missing new line before opening brace.

ArtOfCode- commented on src/WebUI/CodidactOptions.cs:
```diff
@@ -0,0 +1,12 @@
+namespace Codidact.WebUI
+{
+    public class CodidactOptions {
```
Missing new line before opening brace.

ArtOfCode- commented on src/WebUI/CodidactOptions.cs:
```diff
@@ -0,0 +1,12 @@
+
+        public string Hostname { get; set; } = "codidact.com";
+
+        public string CommunitySeperator { get; set; } = "community";
```
Typo: `Separator`

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+
+using Codidact.Domain.Extensions;
+
+namespace Codidact.WebUI {
```
Missing new line

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+using Codidact.Domain.Extensions;
+
+namespace Codidact.WebUI {
+    public class UrlSchemaMiddleware {
```
Missing new line

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            HttpRequest request = context.Request;
+
+            if(options.Value.UseSubdomainSchema) {
```
Missing space, missing new line

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+            IOptionsSnapshot<CodidactOptions> options)
+        {
+            // Deal with naughty users that type the IP address directly.
+            if(!request.Host.Host.EndsWith(options.Value.Hostname)) {
```
Missing space, missing new line

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+            var subdomains = request.Host.Host.TrimSuffix(options.Value.Hostname);
+            var parts = subdomains.Split(".", StringSplitOptions.RemoveEmptyEntries);
+
+            if(parts.Length == 1) {
```
Missing space, missing new line

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+            if(parts.Length == 1) {
+                request.Host = new HostString(options.Value.Hostname);
+                request.Path = $"/community/{parts[0]}{request.Path}";
+            } else if(parts.Length > 1) {
```
Missing new line, missing space, missing new line

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,63 @@
+        private void RewriteCommunitySeperator(HttpRequest request,
+            IOptionsSnapshot<CodidactOptions> options)
+        {
+            if(request.Path.StartsWithSegments($"/{options.Value.CommunitySeperator}")) {
```
Missing space, missing new line

ArtOfCode- commented on tests/WebUI.IntegrationTests/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,75 @@
+{
+    public class UrlSchemaMiddlewareTests
+    {
+        private async Task DoTest(string original, string expected, Action<CodidactOptions> configure) {
```
Missing new line

ArtOfCode- commented on tests/WebUI.IntegrationTests/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,75 @@
+        [InlineData("https://foo.bar.codidact.com", "https://codidact.com/error/404")]
+        [InlineData("https://127.0.0.1", "https://codidact.com/")]
+        [InlineData("https://foo.codidact.com:8080", "https://codidact.com/community/foo/")]
+        public async Task Test_WithSubdomainSchema(string input, string expected) {
```
Missing new line

ArtOfCode- commented on tests/WebUI.IntegrationTests/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,75 @@
+        [InlineData("https://bar.example.com/meta", "https://example.com/meta")]
+        [InlineData("https://127.0.0.1", "https://example.com/")]
+        [InlineData("https://example.com:8080", "https://example.com/")]
+        public async Task Test_WithoutSubdomainSchema(string input, string expected) {
```
Missing new line

ArtOfCode- commented on tests/WebUI.IntegrationTests/UrlSchemaMiddleware.cs:
```diff
@@ -0,0 +1,75 @@
+        [Theory]
+        [InlineData("https://another.co.uk/comunidad/foo", "https://another.co.uk/community/foo")]
+        [InlineData("https://another.co.uk/community/bar/meta", "https://another.co.uk/community/bar/meta")]
+        private async Task Test_CommunitySeperator(string input, string expected) {
```
Missing new line

ArtOfCode- reviewed (changes requested):
There's a few style errors in this PR. I know it's not finished yet, but probably worth pointing them out now so you don't end up doing it all at the end :)

Style guide here: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

asynts commented:
Wow! Quite the effort. Fixed them along a few others that I found while doing so.

ArtOfCode- commented on src/WebUI/UrlSchemaMiddleware.cs:
```diff
@@ -36,17 +39,21 @@ public UrlSchemaMiddleware(RequestDelegate next)
 
-            var subdomains = request.Host.Host.TrimSuffix(options.Value.Hostname);
-            var parts = subdomains.Split(".", StringSplitOptions.RemoveEmptyEntries);
+            string subdomains = request.Host.Host.TrimSuffix(options.Value.Hostname);
```
[You don't _have_ to use explicit types for variables](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions#implicitly-typed-local-variables), you can use `var` if you want to. This is good either way, whatever you choose.

asynts replied:
It's an extension method, at first glance it could be unclear that is returned. I'll leave it like that it shouldn't make much of a difference anyway.

ArtOfCode- commented:
Looks good. Let us know when this is ready to merge :)

asynts committed:
Fixing style errors.

misha130 commented:
I wrote this before can you move the middleware to a folder called middleware? Likewise for test.

Also the homepage tests are broken
Also this is true because of meta:
`        [InlineData("https://codidact.com/foo/meta", "https://codidact.com/community/foo/meta")]
`

asynts committed:
Move middleware into Middlewares folder.

asynts commented:
> I wrote this before can you move the middleware to a folder called middleware? Likewise for test.

I couldn't find that, where did you write that? It is now fixed.

  > Also the homepage tests are broken

The tests pass on my machine and I see no problem with the controller itself. Obviously requests to e.g. `https://codidact.com/community/foo` won't produce a result because that isn't implemented; do you mean that?

  > Also this is true because of meta:
  >
  >     [InlineData("https://codidact.com/foo/meta", "https://codidact.com/community/foo/meta")]

Not sure what you mean by that. The middleware should leave that URL unchanged.

misha130 reviewed (commented):
>I couldn't find that, where did you write that? It is now fixed.

https://github.com/codidact/core/pull/23#discussion_r368755369
OK, I saw it.

> The tests pass on my machine, also I see no problem with the controller itself. Obviously requests to e.g. `https://codidact.com/community/foo` won't produce a result because that isn't implemented; do you mean that?

They failed due migration. Need to add (this is a note for myself for later): 
 ```
          if (env.EnvironmentName != "Test")
            {
                ApplyDatabaseMigrations(app, logger);
            }
```

> Not sure what you mean by that. The middleware should leave that URL unchanged.

But it doesn't. Try it

misha130 commented:
The options should go to the Settings folder as well

asynts commented:
> The options should go to the Settings folder as well

Is there a file hierarchy decided somewhere that I am not aware of?

misha130 commented:
> > The options should go to the Settings folder as well
> 
> Is there a file hierarchy decided somewhere that I am not aware of?

There isnt. Its just common practice. WebUI should only have program & startup on the root.

asynts commented:
I don't know if that is a good idea, adding a new folder for everything. I mean for middlewares it makes sense, but I don't see another settings class pop up soon.

misha130 commented:
Ok, fair enough. We can do a folder called "Common" so as not to add a folder for every little thing.
But since this is the options for the whole project maybe put that common in the infrastructure project rather the WebUI, cause otherwise other projects can't use it.

asynts committed:
Move the configuration into the infrasturcture project.

misha130 commented:
> But it doesn't. Try it

Tried these two: 
```
        [InlineData("https://codidact.com/foo/hi", "https://codidact.com/community/foo/hi")]
        [InlineData("https://codidact.com/foo/meta", "https://codidact.com/community/foo/meta")]
```
They now fail as they should so that is resolved too

asynts commented:
Before there is any more confusion:

  > [...] I'll close #23 and salvage some of the configuration stuff in a separate pull request, unless anyone objects. ([read the whole comment][1])
 
  [1]: https://github.com/codidact/core/issues/22#issuecomment-577171194
